client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -122,10 +153,21 @@ private void handleMissingLog() {
      * message find which new segment it should go to and send it there. This can happen recursively if segments turn
      * over very quickly.
      */
-    @GuardedBy("lock")
     private void handleLogSealed(Segment segment) {
-        List<PendingEvent> toResend = selector.refreshSegmentEventWritersUponSealed(segment);
-        resend(toResend);
+        lock.readLock().lock();
+        /* Using segmentSealedLock the following behaviour is enforced
+            - Prevent concurrent segmentSealedCallback for different segments from being invoked concurrently.
+            - Ensure waiting segmentSealedCallbacks are invoked before the next write is invoked.
+            This ensures that resend() would be invoked again if we observe a segment sealed exception.
+         */
+        segmentSealedLock.lock();
```

Comment:
having locks in succession like this makes the code prone to deadlocks if at some point in the future we end using them in the opposite order. could we rely on a single lock, even if not optimal? you probably have thought about it and thought that it was best like this, in which case, we might want to document somewhere in this class that one needs to be careful about it.